File: Exercism/python/house/house.py
def recite(start_verse, end_verse):
    return [_verse(n) for n in range(start_verse, end_verse + 1)]


def _verse(n):
    things_that_did = " ".join(
        f"the {thing} that {did}" for thing, did in reversed(_THINGS_THAT_DID[:n])
    )
    return f"This is {things_that_did}."


_THINGS_THAT_DID = [
    ("house", "Jack built"),
    ("malt", "lay in"),
    ("rat", "ate"),
    ("cat", "killed"),
    ("dog", "worried"),
    ("cow with the crumpled horn", "tossed"),
    ("maiden all forlorn", "milked"),
    ("man all tattered and torn", "kissed"),
    ("priest all shaven and shorn", "married"),
    ("rooster", "crowed in the morn that woke"),
    ("farmer sowing his corn", "kept"),
    ("horse and the hound and the horn", "belonged to"),
]

# ("morn", "woke"), could be a thing that did...

File: Exercism/haskell/house/src/House.hs
{-# LANGUAGE OverloadedStrings #-}

module House (rhyme) where

import Data.List (intercalate)


data SPO = SPO { subject :: String
               , predicate :: String
               , object :: NP
               } deriving (Show)

data RCl = RCl SPO | End deriving (Show)

data NP = NP { headWord :: String
             , rCl :: RCl
             } deriving (Show)

rhyme :: String
rhyme = intercalate "\n" $ map showSPO verses

verses :: [SPO]
verses = scanl embed initialStory subStories

initialStory :: SPO
initialStory = SPO "This" "is" (NP "house that Jack built" End)

subStories :: [(String, String)]
subStories = [ ("malt", "lay in")
             , ("rat", "ate")
             , ("cat", "killed")
             , ("dog", "worried")
             , ("cow with the crumpled horn", "tossed")
             , ("maiden all forlorn", "milked")
             , ("man all tattered and torn", "kissed")
             , ("priest all shaven and shorn", "married")
             , ("rooster that crowed in the morn", "woke")
             , ("farmer sowing his corn", "kept")
             , ("horse and the hound and the horn", "belonged to")
             ]

makeRCl :: String -> String -> NP -> NP
makeRCl hw verb np = NP hw $ RCl (SPO  "that" verb np)

embed :: SPO -> (String, String) -> SPO
embed (SPO s p o) (hw, verb) = SPO s p (makeRCl hw verb o)

showSPO :: SPO -> String
showSPO (SPO s p o) = unwords [s, p, showNP o]

showNP :: NP -> String
showNP np = "the " ++ headWord np ++ showRCl (rCl np)

showRCl :: RCl -> String
showRCl (RCl spo) = "\n" ++ showSPO spo
showRCl End = ".\n"

